rest/route/keys_routes.go

```diff
@@ -52,3 +67,70 @@ func (h *keysGetHandler) Execute(ctx context.Context, sc data.Connector) (Respon
 		Result: userKeys,
 	}, nil
 }
+
+type keysPostHandler struct {
+	keyName  string
+	keyValue string
+}
+
+func (h *keysPostHandler) Handler() RequestHandler {
+	return &keysPostHandler{}
+}
+
+func (h *keysPostHandler) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	key := model.APIPubKey{}
+	if err := util.ReadJSONInto(body, &key); err != nil {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("failed to unmarshal public key: %s", err),
+		}
+	}
+
+	if err := h.validatePublicKey(key); err != nil {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("invalid public key: %s", err),
+		}
+	}
+
+	return nil
+}
+
+// XXX: If you are changing the validation in this function, you must also
+// update keyBaseValid in directives.spawn.js
+func (h *keysPostHandler) validatePublicKey(key model.APIPubKey) error {
```

Comment:
I think we should check that the key name is not in use by that user

Reply:
AddPublicKey's update selector does this.